### PR TITLE
[Editorial] Catalan numbers

### PR DESCRIPTION
**Pull request by brebenelmihnea**
_Place an "x" in the corresponding checkbox if it is done or does not apply to this pull request._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions).
  - I understand that if it is clear that I have not attempted to follow these conventions, my PR will be closed.
  - If changes are requested, I will re-request a review after addressing them.
- [ ] I have linked this PR to any issues that it closes.

### CODE REVIEW

File: content/4_Gold/Combinatorics.mdx [hunk 1]
```diff
@@ -807,6 +807,102 @@ print()
...
+Consider a convex polygon with $n+2$ sides divided into $n$ triangles by connecting vertices with non-intersecting lines.
+The number of different ways to divide the polygon in this way is equal to $C_n$.
+
+Here's the particular case for $n=3$ in which we have $C_4=$:
+
+![polygons](./assets/catalan-polygons.png)
+
+$C_n$ is also equal to the number of mountain ranges - see [Dyck words](https://en.wikipedia.org/wiki/Dyck_language) - of length $2n$ consisting of $n$ upstrokes and $n$ downstrokes.
+
+![mountains](./assets/catalan-mountain-strokes.png)
```
Comment (SansPapyrus683) [inline content/4_Gold/Combinatorics.mdx:868]: this example seems a bit too niche

File: content/4_Gold/Combinatorics.mdx [hunk 2]
```diff
@@ -807,6 +807,102 @@ print()
...
+For example, there are $C_4=14$ paths in a $4 \times 4 grid$:
+
+![lattice points](./assets/catalan-lattice-paths.png)
+
+The next two examples are a bit more niche, but they're still interesting to think about.
+
+Consider a convex polygon with $n+2$ sides divided into $n$ triangles by connecting vertices with non-intersecting lines.
+The number of different ways to divide the polygon in this way is equal to $C_n$.
+
+Here's the particular case for $n=3$ in which we have $C_4=$:
```
Comment (SansPapyrus683) [inline content/4_Gold/Combinatorics.mdx:862]:
```
> says example is for n=3
> C_4
> what
```

File: content/4_Gold/Combinatorics.mdx [hunk 3]
```diff
@@ -807,6 +807,102 @@ print()
...
+```cpp
+#include <iostream>
+
+using namespace std;
+
+const int MOD = 1e6;
+const int MAX_N = 2e3;
+
+int main() {
+	int comb[MAX_N][MAX_N];
```
Comment (SansPapyrus683) [inline content/4_Gold/Combinatorics.mdx:884]: you'd think a better name would be `catalan`

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: content/4_Gold/Combinatorics.mdx [hunk 4]
```diff
@@ -807,6 +807,101 @@ print()
 </PySection>
 </LanguageSection>
 
+## Catalan Numbers
+
+The [Catalan numbers](https://en.wikipedia.org/wiki/Catalan_number) are a sequence of positive integers that can be very useful in counting problems in combinatorics. Some of the well-known applications are: count certain types of lattice path in a grid, counting binary trees, number of valid paranthesis strings of length $n$, and many more. The $n$-th Catalan can be expressed as following using binomial coeffiecients:
```
Comment (ryanchou-dev) [inline content/4_Gold/Combinatorics.mdx]: one resource i think is great as well is the one by cp-algo. maybe also link it :D

https://cp-algorithms.com/combinatorics/catalan-numbers.html

**Review by ryanchou-dev — APPROVED** [review]
love the drawings!

File: content/4_Gold/Combinatorics.mdx (hunk 2 above)
Comment (brebenelmihnea) [reply content/4_Gold/Combinatorics.mdx:862]: ups

File: content/4_Gold/Combinatorics.mdx (hunk 3 above)
Comment (brebenelmihnea) [reply content/4_Gold/Combinatorics.mdx:884]: done

File: content/4_Gold/Combinatorics.mdx (hunk 1 above)
Comment (brebenelmihnea) [reply content/4_Gold/Combinatorics.mdx:868]: what do you mean?

File: content/4_Gold/Combinatorics.mdx (hunk 4 above)
Comment (brebenelmihnea) [reply content/4_Gold/Combinatorics.mdx]: Done!

File: content/4_Gold/Combinatorics.mdx (hunk 1 above)
Comment (brebenelmihnea) [reply content/4_Gold/Combinatorics.mdx:879]: Idk, it's one of the well-known applications

Comment (SansPapyrus683) [reply content/4_Gold/Combinatorics.mdx:879]: is
is it
@ryanchou-dev thoughts?

Comment (envyaims) [reply content/4_Gold/Combinatorics.mdx:879]: i dont mind it

File: content/4_Gold/Combinatorics.mdx [hunk 5]
```diff
@@ -807,6 +807,122 @@ print()
...
+
+$C_n$ is also equal to the number of mountain ranges - see [Dyck words](https://en.wikipedia.org/wiki/Dyck_language) - of length $2n$ consisting of $n$ upstrokes and $n$ downstrokes.
+
+![mountains](./assets/catalan-mountain-strokes.png)
+
+<FocusProblem problem="sample3" />
+
+### Implementation
+
+**Time Complexity:** $\mathcal{O}(?)$
```
Comment (SansPapyrus683) [inline content/4_Gold/Combinatorics.mdx:890]: fill this in

Comment (brebenelmihnea) [reply content/4_Gold/Combinatorics.mdx]: done

**Comment by brebenelmihnea** [comment]
@SansPapyrus683 merge?

File: content/4_Gold/Combinatorics.mdx (hunk 1 above)
Comment (SansPapyrus683) [reply content/4_Gold/Combinatorics.mdx:884]: unresolved

Comment (ryanchou-dev) [reply content/4_Gold/Combinatorics.mdx:884]: yeah i believe it's pretty well known - however the linked wiki article for dyck words might be a little confusing?

how about this one: https://mathcircle.berkeley.edu/sites/default/files/BMC6/pdf0607/catalan.pdf#page=2

**Comment by bqi343** [comment]
Why was the focus problem merged without an explanation?

ref: https://forum.usaco.guide/t/skyline-problem/6484

**Comment by SansPapyrus683** [comment]
looking at the code it seems the answer is just the nth catalan number...? i thought it was just a template problem lol

**Comment by bqi343** [comment]
> looking at the code it seems the answer is just the nth catalan number...?

yes but why?

**Comment by brebenelmihnea** [comment]
@bqi343 because I don't known how to demonstrate it.

**Comment by bqi343** [comment]
ok ... then it shouldn't be in the module. also, I don't think Catalan numbers are something you need to know for Gold.

maybe once you have a solution explanation you can add it as one of the practice problems in the paths on grids module (since there's a bijection with paths passing below the main diagonal).

**Comment by brebenelmihnea** [comment]
Ok, then tell me where I should move all the stuff or how you wish to proceed.

The [issue](https://forum.usaco.guide/t/skyline-problem/6484) is about the recurrence formula in the code, which I could rewrite more cleanly. Or, I can find another problem with direct application of Catalan numbers, e.g., [Bracket Sequence I](https://cses.fi/problemset/task/2064). My intention with this problem was to illustrate that these numbers do have indeed many applications beyond those mentioned by me.

**Comment by bqi343** [comment]
Given that there's no precedent for Catalan numbers in USACO, you can make a new module in the Advanced section and move everything there.

> The [issue](https://forum.usaco.guide/t/skyline-problem/6484) is about the recurrence formula in the code, which I could rewrite more cleanly.

I think it's obvious that the code computes the Catalan numbers. The issue (IMO) is why the number of good skylines is given by a Catalan number?

> Or, I can find another problem with direct application of Catalan numbers, e.g., [Bracket Sequence I](https://cses.fi/problemset/task/2064).

Sure.

**Comment by brebenelmihnea** [comment]
I think a module just for Catalan numbers is a little too much, but sure. I was thinking of something like `Advanced Combinatorics` encompassing Catalan numbers and the Stars and Bars subsection. That's if you think Stars and Bars could be moved too.

**Comment by bqi343** [comment]
I don't count Stars and Bars as an advanced topic - it's just evaluating a combination.

**Comment by ahmetalp2008** [comment]
I think there is a mistake in Catalan numbers-lattice points picture. 

Bottom-left picture need to be replaced with "RRUURURU".

**Comment by ryanchou-dev** [comment]
you're right, sorry about that.

@brebenelmihnea could you fix it?
i can also do some image editing if you don't have the file anymore, just let me know

**Pull request merged**